mlir/include/mlir/Support/ThreadLocalStore.h

```diff
@@ -0,0 +1,52 @@
+//===- ThreadLocalStore.h - ThreadLocalStore class --------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the ThreadLocalStore class. This class provides support for
+// allocating thread local objects that have a static lifetime.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef MLIR_SUPPORT_THREADLOCALSTORE_H
+#define MLIR_SUPPORT_THREADLOCALSTORE_H
+
+#include "mlir/Support/LLVM.h"
+#include "llvm/Support/Allocator.h"
+#include <utility>
+
+namespace mlir {
+/// This class allocates and stores thread local values of the given value type
+/// with static lifetime. A thread local bump pointer allocator ensures parallel
+/// allocations without any need for further synchronization.
+template <typename ValueT>
+class ThreadLocalStore {
+public:
+  ThreadLocalStore() = default;
+
+  /// Allocates a value type instance for the current thread.
+  template <typename... Args>
+  ValueT *create(Args &&...args) {
```

Comment:
Well here I was considering if the ValueT should be a create function template or a class template?

I went for putting it on the class since that is a better fit for the use case. I want people to only create DistinctAttrs with the distinctAttrStore.